### ISSUE TITLE
[Question] How can I merge several pdf (all images) with the same page size?

### ISSUE THREAD

**Issue by datalw** [pikepdf/pikepdf#169]
Hi everyone,
I am very new to pikepdf. Great work first of all!

I have a question: how can I merge pdfs into one with the same page size, e.g. all in "A4" horizontal format? The small images should remain their original size and large ones may be resized to fit into A4 format. 
I searched on the internet as well as in the documentation, and could only find how to set page size for new pages. I would appreciate any help for this topic :D Thanks a lot!

Right now I am having this kind of problem with e.g. 3 different page sizes: 
![grafik](https://user-images.githubusercontent.com/45690539/107565951-8e5b5d80-6be4-11eb-8c5b-b81525871e54.png)

**Comment by jbarlow83** [comment]
You can use `pikepdf.Page.as_formx_object` to capture an existing page as a Form XObject, which is PDF-speak for a group of drawable things. This way, all of the existing content is grouped together. You would then resize the page by manipulating the size of the MediaBox, and write a new content stream that draws the original content in its new position at a different scale. The PdfMatrix class should help with calculating a matrix to describe the new position.

pikepdf has all the components needed to do this but there isn't a way to do what you describe without digging into the PDF reference manual.

https://pikepdf.readthedocs.io/en/latest/api/models.html#pikepdf.Page.add_resource

https://pikepdf.readthedocs.io/en/latest/api/models.html#pikepdf.Page.as_form_xobject

**Comment by dreua** [comment]
Not sure about your exact use case, but this sounds like it may be much easier to do with a small LaTeX document. Have you considered that? Afaik PDF pages can be included like images. There is even packages like [PyLaTeX](https://jeltef.github.io/PyLaTeX/current/) if you need to fully automate this.

**Comment by jbarlow83** [comment]
True, if you can modify the input before your content gets converted to PDF that is often easier.

**Comment by datalw** [comment]
@jbarlow83 @dreua Thanks a lot for the reply. My exact use case is to generate several figures in my python script, and then put them all in one PDF file. Therefore, I can modify the figure format while saving. Since the figures have different sizes, I would like to have the same page size for all figures, while the figure size remains original - esp. not zomming in the small ones. 
I have barely experience with automating PDF generation... What would be the easiest way to achieve the goal above from your point of view?

**Comment by datalw** [comment]
Since I tried different ways and gathered some experience, I come to make an update for this question. 

Method 1: save the figures first in png, then write them into pdf. For this method, I used the library fpdf.
Advantages: very flexible to set figure position on the page and to recale the figures with a few codes.
Disadvantages: saving and writing process is very slow comparing to the same resolution of figures saved in pdf format.

Method 2: save the figures in pdf, then create blank page in A4 size. Rescale the size of each figure saved in pdf earlier to fit into A4 page while remaining the width-heigh-proportion, then merge it into the blank page. For this method I used PyPDF2.
Advantage: writing speed a little bit faster while the resolution is much higher than method 1. 
Disadvantages: setting position of the figure and recaling are possible, but more codes are needed than method 1.
Bause high resolution and faster writing are very important for me, I chose method 2 for my programm.

Since I didn't test both methods using the same library, the different of the writing speed could also be due to different libraries. Anyway, I hope this information could be useful for others having the some question/needs.
